app/gui2/src/util/ffi.ts

```diff
@@ -0,0 +1,406 @@
+    | { Array: Variant.Array }
+    | { Tuple: Variant.Tuple }
+    | { Annotated: Variant.Annotated }
+    | { AnnotatedBuiltin: Variant.AnnotatedBuiltin }
+    | { Documented: Variant.Documented }
+    | { ConstructorDefinition: Variant.ConstructorDefinition }
+
+  module Variant {
+    export interface Invalid {
+      error: Error
+      ast: Tree
+    }
+
+    export interface BlockLine {
+      newline: Token.Newline
+      expression: Opt<Tree>
+    }
+
+    export interface BlockOperatorLine {}
+
+    type OperatorOrError = { Ok: Token } | { Err: Error }
+    export interface BodyBlock {
+      statements: BlockLine[]
+    }
+
+    export interface ArgumentBlockApplication {
+      lhs: Opt<Tree>
+      arguments: BlockLine[]
+    }
+    export interface OperatorBlockApplication {
+      lhs: Opt<Tree>
+      expressions: BlockOperatorLine[]
+      excess: BlockLine[]
+    }
+    export interface Ident {
+      token: Token.Ident
+    }
+
+    export interface FractionalDigits {
+      dot: Token.Operator
+      digits: Token.Digits
+    }
+
+    export interface Number {
+      base: Opt<Token.NumberBase>
+      integer: Opt<Token.Digits>
+      fractional_digits: Opt<FractionalDigits>
+    }
+    export interface Wildcard {
+      token: Token.Wildcard
+      de_bruijn_index: Opt<number>
+    }
+    export interface AutoScope {
+      token: Token.AutoScope
+    }
+
+    export interface TextLiteral {
+      open: Opt<Token.TextStart>
+      newline: Opt<Token.Newline>
+      elements: TextElement[]
+      close: Opt<Token.TextEnd>
+    }
+    export interface App {
+      func: Tree
+      arg: Tree
+    }
+    export interface NamedApp {
+      func: Tree
+      open: Opt<Token.OpenSymbol>
+      name: Token.Ident
+      equals: Token.Operator
+      arg: Tree
+      close: Opt<Token.CloseSymbol>
+    }
+    export interface DefaultApp {
+      func: Tree
+      default: Token.Ident
+    }
+    export interface OprApp {
+      lhs: Opt<Tree>
+      opr: OperatorOrError
+      rhs: Opt<Tree>
+    }
+    export interface UnaryOprApp {
+      opr: Token.Operator
+      rhs: Opt<Tree>
+    }
+    export interface OprSectionBoundary {
+      arguments: number
+      ast: Tree
+    }
+    export interface TemplateFunction {
+      arguments: number
+      ast: Tree
+    }
+
+    export interface MultiSegmentApp {
+      segments: NonEmptyArray<MultiSegmentAppSegment>
+    }
+
+    export interface MultiSegmentAppSegment {
+      header: Token
+      body: Opt<Tree>
+    }
+
+    export interface ArgumentType {
+      operator: Token.Operator
+      type: Tree
+    }
+
+    export interface ArgumentDefault {
+      equals: Token.Operator
+      expression: Tree
+    }
+
+    export interface ArgumentDefinition {
+      open: Opt<Token.OpenSymbol>
+      open2: Opt<Token.OpenSymbol>
+      suspension: Opt<Token.Operator>
+      pattern: Tree
+      type: Opt<ArgumentType>
+      close2: Opt<Token.CloseSymbol>
```

Comment:
weird ordering? i'd expect `close2` to be next to `close`. very minor nitpick though

Reply:
I kept the ordering identical to the one from the rust codebase. I think it is specified in order that the tokens are present in the source. I'll copy over the doc comments as well for this struct, because it is indeed quite confusing.